crates/matrix-sdk/src/room/mod.rs

```diff
@@ -368,6 +368,71 @@ impl Room {
         Ok(())
     }
 
+    /// Leave this room and all predecessors.
+    /// If any room was in [`RoomState::Invited`] state, it'll also be forgotten
+    /// automatically.
+    ///
+    /// Only invited and joined rooms can be left.
+    pub async fn leave(&self) -> Result<()> {
+        let mut rooms: Vec<Room> = vec![self.clone()];
+        let mut cur_room = self;
+
+        'leave_aggregate_loop: loop {
+            let pred = cur_room.predecessor_room();
+            let pred = pred.map(|this_pred| {
+                cur_room
+                    .client
+                    .get_room(&this_pred.room_id)
+                    .ok_or(Error::PredecessorWrongOrForgotten)
+            });
+
+            match pred {
+                None => break 'leave_aggregate_loop,
+                Some(room) => match room {
+                    Ok(room) => {
+                        rooms.push(room);
+                        cur_room = rooms.last().expect("rooms just pushed so can't be empty");
+                    }
+                    Err(e) => {
+                        debug!("Tombstone following terminated early: {e:?}");
+                        break 'leave_aggregate_loop;
+                    }
+                },
+            }
+        }
+
+        let mut rooms_iter = rooms.into_iter();
+        let batch_size = 5;
+
+        'leave_action_loop: loop {
+            let batch: Vec<Room> = rooms_iter.by_ref().take(batch_size).collect();
+            if batch.is_empty() {
+                break 'leave_action_loop;
+            }
+
+            let batch = batch
+                .iter()
+                .map(|room| match room.state() {
+                    RoomState::Joined => Ok(Some(room.leave_this())),
+                    RoomState::Invited => Ok(Some(room.leave_this())),
+                    RoomState::Knocked => Ok(Some(room.leave_this())),
```

Comment:
Seems like `join_all!()` turned out to be the wrong pattern here.

Using the `buffered()` stream extension seems to produce more readable results:

```rust
let futures: Vec<_> = rooms
    .into_iter()
    .filter_map(|room| match room.state() {
        RoomState::Joined | RoomState::Invited | RoomState::Knocked => {
            Some(async move { room.leave_this().await })
        }
        RoomState::Left | RoomState::Banned => None,
    })
    .collect();

let mut stream = futures_util::stream::iter(futures).buffered(BATCH_SIZE);

for result in stream.next().await {
    match result {
        Ok(_) => todo!(),
        Err(_) => todo!(),
    }
}
```